src/python/pants/backend/docker/rules_test.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from hashlib import sha256
+from os import path
+
+import pytest
+
+from pants.backend.docker.rules import DockerFieldSet, build_docker_image
+from pants.backend.docker.rules_binary import DockerBinary, DockerBinaryRequest
+from pants.backend.docker.rules_context import (
+    DockerBuildContext,
+    DockerBuildContextRequest,
+    create_docker_build_context,
+)
+from pants.backend.docker.target_types import DockerImage
+from pants.core.goals.package import BuiltPackage, PackageFieldSet
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.fs import (
+    EMPTY_DIGEST,
+    EMPTY_FILE_DIGEST,
+    EMPTY_SNAPSHOT,
+    AddPrefix,
+    Digest,
+    MergeDigests,
+    Snapshot,
+)
+from pants.engine.process import Process, ProcessResult
+from pants.engine.target import (
+    DependenciesRequest,
+    FieldSetsPerTarget,
+    FieldSetsPerTargetRequest,
+    Targets,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionMembership
+from pants.testutil.rule_runner import MockGet, run_rule_with_mocks
+from pants.util.ordered_set import FrozenOrderedSet
+
+
+def test_docker_binary_build_image():
+    source_path = "src/test/repo"
+    docker_path = "/bin/docker"
+    dockerfile = path.join(source_path, "Dockerfile")
+    docker = DockerBinary(docker_path)
+    digest = Digest(sha256().hexdigest(), 123)
+    tag = "test:latest"
+    build_request = docker.build_image(tag, digest, source_path, dockerfile)
+
+    assert build_request == Process(
+        argv=(docker_path, "build", "-t", tag, "-f", dockerfile, source_path),
+        input_digest=digest,
+        description=f"Building docker image {tag}",
+    )
+
+
+@pytest.mark.parametrize(
+    "target_values, expected_features",
+    [
+        (
+            dict(),
+            dict(
+                context_root="docker/test/.",
+            ),
+        ),
+        (
+            dict(
+                version="1.2.3",
+            ),
+            dict(
+                version="1.2.3",
+            ),
+        ),
+        (
+            dict(
+                context_root="/",
+            ),
+            dict(
+                context_root=".",
```

Comment:
Thinking we would probably want to test the context root ft in isolation, and not in this test..